src/java/org/apache/cassandra/dht/NormalizedRanges.java

```diff
@@ -0,0 +1,295 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.dht;
+
+import java.util.AbstractList;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.RandomAccess;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterators;
+import com.google.common.collect.PeekingIterator;
+
+import static com.google.common.base.Preconditions.checkState;
+
+/*
+ * Immutable list of ranges that are statically known to be normalized
+ */
+public class NormalizedRanges<T extends RingPosition<T>> extends AbstractList<Range<T>> implements List<Range<T>>, RandomAccess
+{
+    private static final NormalizedRanges EMPTY_NORMALIZED_RANGES = new NormalizedRanges(Collections.emptyList());
+
+    private static final Comparator NORMALIZED_TOKEN_RANGE_COMPARATOR = (o1, o2) -> {
+        Range range = (Range) o1;
+        RingPosition key = (RingPosition) o2;
+        boolean rangeRightIsMin = range.right.isMinimum();
+        boolean keyIsMinimum = key.isMinimum();
+
+        if (keyIsMinimum & rangeRightIsMin)
+            return 0;
```

Comment:
Is this correct? The minimum token is less than everything except only in cases where it represents the upper bound (in which case it means more than everything). Since this is a key being compared to, this seems potentially wrong?

Reply:
Ah, the comment below explains the logic. We can only include the minimum token if we wrap around. I'm not 100% convinced by this though - can we logically ever contain the minimum token? It's less than everything, so even if we wrap around we won't include it.

Reply:
I'll remove it and run the tests. I can't recall what drove the addition of that logic, but I think it causes problems with sorting and binary searching if you don't define it properly/consistently.

How `Range` handles this is not even very consistent in terms of `contains`. If it's a wrap around range then it will include `MIN` but if it's not a wrap around range it will actually include `MIN` in the range `MIN,0`!

I am not sure there is a "correct" behavior here, but the one that maps to the inclusivity behavior of ranges at least seems more intuitive.